### ISSUE TITLE
Waterfox 68 support

### ISSUE THREAD

**Issue by RaitaroH** [akhodakivskiy/VimFx#933]
If there is ONE extension that it makes it impossible to move to anything else besides good old waterfox is vimfx. Could it be possible for this addon to be fully supported under waterfox 68? One thing that I hate is that in new firefox and in chrome vim like extensions don't work on the newtab, don't work in any browser page, can't really focus the url bar and have some urlbar like thing etc. Not to mention they feel clunky.

Documentation ![here](https://github.com/MrAlex94/Waterfox/wiki).

**Comment by lydell** [comment]
Hi! Wow, I didn't know people still used this extension!

If you want Waterfox support, you either have to do it yourself, or hope that somebody else will step up and do it :)

I'll be around if you have any questions about the code base, but as I use neither Waterfox nor VimFX I won't be doing any coding myself.

**Comment by girst** [comment]
I'm still stuck on ff64 and would pitch in some money if this extension 
gets ported to either webextension-experiments (firefox compatible) or 
the waterfox thingie.

**Comment by girst** [comment]
I had a look at porting this to Waterfox 68, and it appears doable. I've installed the 68.0a1 in a VM and after not too many changes (mostly `require`ing some stuff) I managed to get it to run.

There's one caveat: clicking does not work (same as firefox 63+): the nsIDOMWindowUtils interface wasd removed from nsiDOMChromeWindow. This makes the following code fail:

```coffeescript
# extension/lib/utils.coffee:336
      window
        .QueryInterface(Ci.nsIInterfaceRequestor)    # returns a ChromeWindow
        .getInterface(Ci.nsIDOMWindowUtils)          # fails with NS_NOINTERFACE
        .dispatchDOMEventViaPresShell(element, mouseEvent, true)
```

@lydell, as the one most familiar with vimfx' internals: do you know if this is work-around-able? is there another mozilla-internal API to do this, or another way to access the dispatchDOMEventViaPresShell function (it's still there, just not throgh ChromeWindow)?

**Comment by lydell** [comment]
Sorry, I don't know how to dispatch a trusted event if that API is gone/moved. The only thing I can suggest is always using `element.dispatchEvent(mouseEvent)` until you find something else. The only downside is that you won't be able to click on some Fullscreen and Copy buttons. The workaround to that is usually first trying to click with VimFx, then pressing Enter to activate the now focused button.

**Comment by girst** [comment]
I figured that would be the case. Thanks anyways, lydell!

I'm uploading yesterday's patches, before i lose them. I'm down to three exceptions, noted below:

```javascript
//lib/utils.coffee:338:        
    .getInterface(Ci.nsIDOMWindowUtils) //throws NS_NOINTERFACE: "Component does not have requested interface"
    //this interface was removed from chromewindow in ff62; possible workarounds: use e.dispatchEvent(), forward-port to waterfox68
//lib/events.coffee:172:    
    activeElement = utils.getActiveElement(@window) //throws TypeError: "utils is null"
    //not yet debugged
//lib/commands.coffee:1054:  
    developerToolbar = gDevToolsBrowser.getDeveloperToolbar(vim.window) //throws TypeError: "getDeveloperToolbar not a function"
    //breaks escape; getDevToolbar() was removed, need to look into workarounds, otherwise; see bugzil.la/1461970
```
[vimfx.patch.txt](https://github.com/akhodakivskiy/VimFx/files/3169979/vimfx.patch.txt)

**Comment by lydell** [comment]
The developer toolbar was removed entirely from Firefox a while back, wasn't it? If so, you could remove all code and commands related to it.

**Comment by matsuhav** [comment]
[My fork](https://github.com/matsuhav/VimFx) could help you @girst .  I had fixed some code so that "follow link" command works.  It seems it is still working in waterfox 56.2.9. Try it from release.

**Comment by girst** [comment]
@ matsuhav: you are a goddamn saint! `window.windowUtils` instead of `.getInterface(Ci.nsIDOMWindowUtils)` did the trick!

I've also found that the "utils is null" TypeError happens when you click a doorhanger, so that's not too bad. 

I think I'll rebase my changes on to of your (matsuhav's) fork, so they work in firefox/waterfox 56, firefox 64 and waterfox 68, and maybe open a PR then.

**Comment by girst** [comment]
so, this is now done! I've tested it with firefox-56.0.2-1.fc27, firefox-64.02-1.fc28, waterfox 56 (from waterfox.net) and waterfox 68 (from the 'releases' page on github)

would be great if this was merged :)

**Comment by lydell** [comment]
Fixed in [v0.21.0](https://github.com/akhodakivskiy/VimFx/releases/tag/v0.21.0).
